src/python/pants/backend/go/module.py

```diff
@@ -0,0 +1,203 @@
+class ResolveGoModuleRequest:
+    address: Address
+
+
+# Perform a minimal parsing of go.mod for the `module` and `go` directives. Full resolution of go.mod is left to
+# the go toolchain. This could also probably be replaced by a go shim to make use of:
+# https://pkg.go.dev/golang.org/x/mod/modfile
+def basic_parse_go_mod(raw_text: bytes) -> Tuple[Optional[str], Optional[str]]:
+    module_path = None
+    minimum_go_version = None
+    for line in raw_text.decode("utf-8").splitlines():
+        parts = line.strip().split()
+        if len(parts) >= 2:
+            if parts[0] == "module":
+                if module_path is not None:
+                    raise ValueError("Multiple `module` directives found in go.mod file.")
+                module_path = parts[1]
+            elif parts[0] == "go":
+                if minimum_go_version is not None:
+                    raise ValueError("Multiple `go` directives found in go.mod file.")
+                minimum_go_version = parts[1]
+    return module_path, minimum_go_version
+
+
+# Parse the output of `go mod download` into a list of module descriptors.
+def parse_module_descriptors(raw_json: bytes) -> List[ModuleDescriptor]:
+    module_descriptors = []
+    for raw_module_descriptor in ijson.items(raw_json, "", multiple_values=True):
+        module_descriptor = ModuleDescriptor(
+            import_path=raw_module_descriptor["Path"],
+            module_path=raw_module_descriptor["Path"],
+            module_version=raw_module_descriptor["Version"],
+        )
+        module_descriptors.append(module_descriptor)
+    return module_descriptors
+
+
+@rule
+async def resolve_go_module(
+    request: ResolveGoModuleRequest,
+    goroot: GoLangDistribution,
+    platform: Platform,
+    bash: BashBinary,
+) -> ResolvedGoModule:
+    downloaded_goroot = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        goroot.get_request(platform),
+    )
+
+    targets = await Get(UnexpandedTargets, Addresses([request.address]))
+    if not targets:
+        raise ValueError(f"Address `{request.address}` did not resolve to any targets.")
+    elif len(targets) > 1:
+        raise ValueError(f"Address `{request.address}` resolved to multiple targets.")
+    target = targets[0]
+
+    sources = await Get(SourceFiles, SourceFilesRequest([target.get(GoModuleSources)]))
+    flattened_sources_digest = await Get(
+        Digest, RemovePrefix(sources.snapshot.digest, request.address.spec_path)
+    )
+    flattened_sources_snapshot = await Get(Snapshot, Digest, flattened_sources_digest)
+    if (
+        len(flattened_sources_snapshot.files) not in (1, 2)
+        or "go.mod" not in flattened_sources_snapshot.files
+    ):
+        raise ValueError(f"Incomplete go_module sources: files={flattened_sources_snapshot.files}")
+
+    analyze_script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "analyze.sh",
```

Comment:
I think you could directly express this via a `Process`, unless I'm missing something. Use `env` to hardcode those 3 env vars. Use `await Get(Digest, CreateDigest([Directory()])` to recreate `mkdir`.

Reply:
I can't. The `go` demands an absolute path for GOPATH and that can only be obtained from within the sandbox.

Reply:
I switched to using a bash script because I had tried to set relative paths for those variables via `env`.

Reply:
Sg. A comment would be helpful motivating that

Reply:
Also, I recommend moving everything you can via a `Process`, meaning everything but those two env vars and the `exec`. For example, right now you are assuming `mkdir` exists on the machine, which is one more implicit runtime dep we'd rather avoid.

Reply:
I'll add a comment.